experimental/casemapping/src/internals.rs

```diff
@@ -0,0 +1,1066 @@
+        let mut lengths = self.slot_value(base_idx, CaseMappingExceptionSlot::FullMappings);
+        let mut data_idx = self.slot_index(base_idx, CaseMappingExceptionSlot::FullMappings) + 1;
+
+        // Skip past previous slots
+        for _ in 0..(slot as usize) {
+            let str_len = lengths & Self::FULL_MAPPINGS_LENGTH_MASK;
+            data_idx += str_len as usize;
+            lengths >>= Self::FULL_MAPPINGS_LENGTH_SHIFT;
+        }
+
+        // Return range of this string
+        let str_len = (lengths & Self::FULL_MAPPINGS_LENGTH_MASK) as usize;
+        data_idx..(data_idx + str_len)
+    }
+
+    // Full mapping strings are stored as UTF16 immediately following the full mappings slot.
+    // The length of each full mapping string (in code units) is stored in a nibble of the slot.
+    fn full_mapping_string(
+        &self,
+        base_idx: u16,
+        slot: FullMapping,
+    ) -> Result<String, DecodeUtf16Error> {
+        debug_assert!(self.has_slot(base_idx, CaseMappingExceptionSlot::FullMappings));
+        let range = self.full_mapping_string_range(base_idx, slot);
+        let iter = self.raw.as_slice()[range]
+            .iter()
+            .map(|&ule| <u16 as AsULE>::from_unaligned(ule));
+        char::decode_utf16(iter).collect()
+    }
+
+    // Given a base index, returns the length of the closure string
+    fn closure_len(&self, base_idx: u16) -> usize {
+        debug_assert!(self.has_slot(base_idx, CaseMappingExceptionSlot::Closure));
+        let value = self.slot_value(base_idx, CaseMappingExceptionSlot::Closure);
+        (value & Self::CLOSURE_MAX_LENGTH) as usize
+    }
+
+    // The closure string is stored as UTF16 following the full mappings (if they exist) or the
+    // closure slot (if no full mappings are present). The closure slot stores the length of the
+    // closure string (in code points).
+    // This function returns an iterator over the characters of the closure string.
+    fn closure_string(&self, base_idx: u16) -> Result<String, DecodeUtf16Error> {
+        let start_idx = if self.has_slot(base_idx, CaseMappingExceptionSlot::FullMappings) {
+            self.full_mapping_string_range(base_idx, FullMapping::Title)
+                .end
+        } else {
+            self.slot_index(base_idx, CaseMappingExceptionSlot::Closure) + 1
+        };
+        let closure_len = self.closure_len(base_idx);
+        let u16_iter = self.raw.as_slice()[start_idx..]
+            .iter()
+            .map(|&ule| <u16 as AsULE>::from_unaligned(ule));
+        char::decode_utf16(u16_iter).take(closure_len).collect()
+    }
+
+    fn add_full_and_closure_mappings<S: SetAdder>(&self, base_idx: u16, set: &mut S) {
+        if self.has_slot(base_idx, CaseMappingExceptionSlot::FullMappings) {
+            let mapping_string = self
+                .full_mapping_string(base_idx, FullMapping::Fold)
+                .expect("No unpaired surrogates");
```

Comment:
Issue: No panics in runtime code (unless you can prove they are unreachable).

I assume this is runtime code, but if this is data builder code, you should hide it behind a flag such as `provider_transform_internals` like we do elsewhere.

Reply:
(This is runtime code.)

The design here is that `try_from_icu` is the only way to build a CaseMapping. In that function, we call the same helper function on every full mapping string and propagate errors. It is therefore impossible (short of writing broken serialized data by hand and deserializing it) for this to fail.

I will change the expect message here to something like "Data validated on construction".

Reply:
`try_from_icu` runs in the data transformer, but the class could also be constructed from serialized postcard data, correct? If so, the current policy is no panicking for code reachable by malformed/malicious data.

Reply:
Related: https://github.com/unicode-org/icu4x/issues/1183

Reply:
(this is still unresolved and I'm unsure how to move forward)